File: go/src/pr048/pr048.go
//////////////////////////////////////////////////////////////////////
// Problem 48
//
// 18 July 2003
//
// The series, 1^1 + 2^2 + 3^3 + ... + 10^10 = 10405071317.
//
// Find the last ten digits of the series, 1^1 + 2^2 + 3^3 + ... + 1000
// ^1000.
//
// 9110846700
//////////////////////////////////////////////////////////////////////

package pr048

import (
	"fmt"
)

const modulus = 10000000000

func Run() {
	sum := int64(0)
	for i := 1; i <= 1000; i++ {
		sum = (sum + expt(int64(i), i)) % modulus
	}

	fmt.Printf("%d\n", sum)
}

func expt(base int64, power int) (result int64) {
	result = 1
	for i := 0; i < power; i++ {
		result = (result * base) % modulus
	}
	return
}

File: rust/src/pr048.rs
// Problem 48
//
// 18 July 2003
//
//
// The series, 1^1 + 2^2 + 3^3 + ... + 10^10 = 10405071317.
//
// Find the last ten digits of the series, 1^1 + 2^2 + 3^3 + ... + 1000^1000.

define_problem!(pr048, 48, 9110846700);

const MODULUS: u64 = 10000000000;

fn pr048() -> u64 {
    let mut sum = 0;
    for i in 1 .. 1001 {
        sum = (sum + expt(i, i)) % MODULUS;
    }

    sum
}

fn expt(base: u64, power: u64) -> u64 {
    let mut result = 1;
    for _ in 0 .. power {
        result = (result * base) % MODULUS;
    }
    result
}
